packages/@react-stately/table/src/TableUtils.ts

```diff
@@ -133,42 +65,215 @@ export interface IColumn {
+    let baseSize = 0;
+    let flex = 0;
     if (isStatic(width)) {
-      let w = parseStaticWidth(width, availableWidth);
-      w = Math.max(
-        getMinWidth(column.minWidth, availableWidth),
-        Math.min(Math.floor(w), getMaxWidth(column.maxWidth, availableWidth)));
-      acc.staticColumns.push({index, column, width: w} as IndexedColumn);
-      acc.dynamicColumns.push(null);
-      remainingSpace -= w;
+      baseSize = parseStaticWidth(width, availableWidth);
+      frozen = true;
     } else {
-      let w = parseFractionalUnit(width);
-      acc.staticColumns.push(null);
-      acc.dynamicColumns.push({index, column, width: w} as IndexedColumn);
+      flex = parseFractionalUnit(width);
+      if (flex <= 0) {
+        frozen = true;
+      }
     }
-    return acc;
-  }, {staticColumns: [] as IndexedColumn[], dynamicColumns: [] as IndexedColumn[]});
-  let newColWidths = getDynamicColumnWidths(dynamicColumns, remainingSpace, availableWidth);
 
-  return staticColumns.map((col, i) => {
-    if (col) {
-      return col.width;
+    let min = getMinWidth(column.minWidth ?? getDefaultMinWidth?.(index) ?? 0, availableWidth);
+    let max = getMaxWidth(column.maxWidth, availableWidth);
+    let hypotheticalMainSize = Math.max(min, Math.min(baseSize, max));
+    return {
+      frozen,
+      baseSize,
+      hypotheticalMainSize,
+      min,
+      max,
+      flex,
+      targetMainSize: null,
+      violation: ''
+    };
+  });
+  // 9.7.1
+  // We don't make use of flex basis, it's always 0, so we are always in 'grow' mode.
+  // 9.7.2
+  flexItems.forEach(item => {
+    if (item.frozen) {
+      item.targetMainSize = item.hypotheticalMainSize;
+      return;
+    }
+    if (item.baseSize > item.hypotheticalMainSize) {
+      item.frozen = true;
+      item.targetMainSize = item.hypotheticalMainSize;
     }
-    return newColWidths[i].width;
   });
+
+  // 9.7.3
+  let initialFreeSpace = availableWidth - flexItems.reduce((acc, item) => {
+    if (item.frozen) {
+      return acc + item.targetMainSize;
+    }
+    return acc + item.baseSize;
+  }, 0);
+
+  let remainingFreeSpace = initialFreeSpace;
+  // 9.7.4
+  // 9.7.4.a
+  while (flexItems.some(item => !item.frozen)) {
+    // 9.7.4.b
+    /**
+     * Calculate the remaining free space as for initial free space,
+     * above (9.7.3). If the sum of the unfrozen flex items’ flex factors is
+     * less than one, multiply the initial free space by this sum (of flex factors).
+     * If the magnitude of this value is less than the magnitude of
+     * the remaining free space, use this as the remaining free space.
+     */
+    remainingFreeSpace = availableWidth - flexItems.reduce((acc, item) => {
+      if (item.frozen) {
+        return acc + item.targetMainSize;
+      }
+      return acc + item.baseSize;
+    }, 0);
+    let flexFactors = flexItems.reduce((acc, item) => {
+      if (!item.frozen) {
+        return acc + item.flex;
+      }
+      return acc;
+    }, 0);
```

Comment:
You could combine the above two loops into a single pass. Might be easier to read without reduce anyway.